Added Alias method

cekefun opened the pull request:
Took so long because of stupid death test which still don't work.

cekefun committed:
Merge pull request #1 from flu-plus-plus/master

fork update

cekefun committed:
Merge pull request #2 from flu-plus-plus/master

merge

cekefun committed:
Added Alias Distribution Method

cekefun committed:
Merge pull request #3 from flu-plus-plus/master

Chart on my fork

cekefun committed:
Alias Method added

cekefun committed:
deleted old Alias

cekefun committed:
Merge branch 'master' of https://github.com/flu-plus-plus/bachelorproef into flu-plus-plus-master

cekefun committed:
Merge branch 'flu-plus-plus-master'

cekefun committed:
Tests aangepast

jonathanvdc commented on src/main/cpp/alias/Alias.h:
```diff
@@ -0,0 +1,35 @@
+ *      Author: cedric
+ */
+#ifndef ALIAS_H_
+#define ALIAS_H_
```
Can you make the include guard identifier conform to Stride's format? That would be:

```c
#ifndef ALIAS_H_INCLUDED
#define ALIAS_H_INCLUDED

...
#endif
```
```

cekefun replied:
This has beem edited out.

jonathanvdc commented on src/main/cpp/alias/Alias.h:
```diff
@@ -0,0 +1,35 @@
+class Alias {
+public:
+	///Constructor. Uses a random seed
+	Alias(const std::vector<double> probabilities);
```
Slapping just `const` on a parameter doesn't *actually* mean anything.

Did you mean to write `const std::vector<double>&`? That's also a whole lot more efficient because it avoids copying the whole `std::vector<double>` on every call to `Alias::Alias`.

cekefun replied:
I forgot to remove that const.

jonathanvdc commented on src/main/cpp/util/AliasUtil.h:
```diff
@@ -0,0 +1,29 @@
+ *      Author: cedric
+ */
+
+#ifndef ALIASUTIL_H_
```
Can you make the include guard identifier conform to Stride's format? That would be:

#ifndef ALIAS_UTIL_H_INCLUDED
#define ALIAS_UTIL_H_INCLUDED

...
#endif

cekefun replied:
The INCLUDED has been added

jonathanvdc commented on src/main/cpp/alias/Alias.cpp:
```diff
@@ -0,0 +1,82 @@
+namespace stride{
+namespace alias{
+
+Alias::Alias(std::vector<double> probabilities) : Alias(probabilities,std::random_device()()) {}
```
I think you need to `#include <random>` before you can use `std::random_device`.

cekefun replied:
Everything to do with random has been removed.

jonathanvdc commented on src/test/cpp/gtester/AliasTest.cpp:
```diff
@@ -0,0 +1,24 @@
+
+TEST(AliasDeathTest, Alias)
+{	
+	//Ruzie met death
```
This might be because calls to `assert` are elided when the application is compiled in release mode, which is like, always, I guess.

jonathanvdc replied:
You can probably work around this by throwing an `EmptyProbabilityException` instance and then catching that with `ASSERT_THROW`. I think one of Lynn's tests does just that. You might want to take a look at it.

jonathanvdc replied:
Also, Dutch comment. Boo!

cekefun replied:
Now it uses throws. Also, the comment has been removed.

jonathanvdc reviewed (changes requested):
Looks pretty good overall. General remark: can you format the code files you created? Use `clang-format` from the command-line, or use an IDE of your choice to call `clang-format`.

cekefun committed:
Changed Include guards
Added std::random
Fixed tests by making it exceptions

cekefun committed:
-Deleted constructors
-Made static function which calls constructor
-Util::Random is an argument instead of a seed
-Formatted

jonathanvdc commented on src/main/cpp/alias/Alias.cpp:
```diff
@@ -0,0 +1,77 @@
+namespace stride {
+namespace alias {
+
+Alias& Alias::CreateDistribution(std::vector<double> probabilities, util::Random& rng)
```
I appreciate the refactor, but you can't just return a reference to a `static` value here. That's super dangerous!

jonathanvdc replied:
Please change this to a regular `Alias` value that is constructed by this function and use a move constructor to return it to the caller.

cekefun replied:
Should it be return *(new Alias(alias,prob,rng))? I don't know any other way make a static function return a reference in c++.

jonathanvdc replied:
Hang on. I'll post a gist of how to solve this problem in just a sec.

jonathanvdc replied:
[There ya go.](https://gist.github.com/jonathanvdc/faa38c7a8320c125d04d9d4adb756ce2) The trick's to rely on a move constructor.

cekefun replied:
sorry, saw this too late. "solved" it another way. I will look at the differences tomorow.

cekefun committed:
made the random number generator a reference.

jonathanvdc commented on src/main/cpp/alias/Alias.h:
```diff
@@ -0,0 +1,46 @@
+	Alias(const Alias&) = delete;
+	/// No assignement operator needed
+	Alias& operator=(const Alias&) = delete;
+	/// Creates an Alias object. The constructor would be way too long.
```
You might want to leave out the "constructor would be way too long." part. Comments that start with three slashes are documentation comments, and gory implementation details aren't all that relevant to the documentation.

jonathanvdc replied:
This is just a minor nitpick, though. Everything else looks absolutely fine.

cekefun committed:
edited comment

cekefun committed:
made it a move
Couldn't make it a reference, will try to solve that tomorrow

jonathanvdc reviewed (approved):
Lgtm
